File: solutions/037/037.py
"""
Project Euler - Problem Solution 037
Problem Title - Truncatable primes
Copyright (c) Justin McGettigan. All rights reserved.
https://github.com/jwmcgettigan/project-euler-solutions
"""

def truncated_nums(prime):
  ''' Generates every truncation of a number. '''
  digits = str(prime)
  for i in range(1, len(digits)):
    yield int(digits[i:])
    yield int(digits[:-i])

def truncatable_primes():
  ''' Finds the sum of the only eleven primes that are
  both truncatable from left to right and right to left. '''
  limit = 1000
  primes, composites = set(), set()
  t_primes = set()
  not_t_primes = {2, 3, 5, 7}

  while len(t_primes) < 11:
    limit *= 2
    for i in range(2, limit+1):
      if i not in composites:
        primes.add(i)
        for j in range(i*i, limit+1, i):
          composites.add(j)
        # ------------------------------------
        truncatable = True
        for num in truncated_nums(i):
          if num not in primes:
            truncatable = False
            break
        if truncatable and i not in not_t_primes:
          t_primes.add(i)
  
  return sum(t_primes)

if __name__ == "__main__":
	print(truncatable_primes())

File: solutions/037/037.js
/**
 * Project Euler - Problem Solution 037
 * Problem Title - Truncatable primes
 * Copyright (c) Justin McGettigan. All rights reserved.
 * https://github.com/jwmcgettigan/project-euler-solutions
 */

/** Return the sum of all elements in an iterable of numbers. */
const sum = (iterable) => {
  let s = 0;
  iterable.forEach((num) => { s += num; });
  return s;
}

/** Generates every truncation of a number. */
function* truncatedNums(prime) {
  const digits = String(prime);
  for(let i = 1; i < digits.length; i++) {
    yield Number(digits.substring(i));
    yield Number(digits.substring(0, digits.length-i));
  }
};

/** Finds the sum of the only eleven primes that are
 * both truncatable from left to right and right to left. */
const truncatablePrimes = () => {
  let limit = 1000;
  const primes = new Set(), composites = new Set();
  const tPrimes = new Set(), not_tPrimes = new Set([2, 3, 5, 7]);

  while(tPrimes.size < 11) {
    limit *= 2;
    
    for(let i = 2; i < limit+1; i++) {
      if(!composites.has(i)) {
        primes.add(i);
        for(let j = i*i; j < limit+1; j += i) {
          composites.add(j);
        }
        // -------------------------------------
        let truncatable = true;
        for(let num of truncatedNums(i)) {
          if(!primes.has(num)) {
            truncatable = false;
            break;
          }
        }
        if(truncatable && !not_tPrimes.has(i)) {
          tPrimes.add(i);
        }
      }
    }
  }

  return sum(tPrimes);
};

console.log(truncatablePrimes());